### PR TITLE
add hw15

### PR DESCRIPTION
**Pull request by IzyNP**

### CODE REVIEW

File: src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package com.izmaylov.hw15.practiceSlide20;
+
+import java.util.NoSuchElementException;
+
+public class MyList<E> {
+        private MyNode<E> head;
+        private MyNode<E> tail;
+        int size = 0;
```
Comment (Fabolo12) [inline src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java]: private

File: src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+
+        public void add(E value){
+            MyNode<E> node = new MyNode<>(value,null, null);
```
Comment (Fabolo12) [inline src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java]: Why does the constructor need 3 parameters, 2 of which are always null

File: src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            if(value == null){
+                throw new NullPointerException();
+            }else {
```
Comment (Fabolo12) [inline src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java]: If you add this check, then before creating a node

File: src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                tail = node;
+                size++;
+            }
+        }
+
+
+        public E getFromLast(E value){
+            if(tail == null){
+                throw new NoSuchElementException();
+            }
```
Comment (Fabolo12) [inline src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java]: not sure if this is a good decision

File: src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            if(last.getValue().equals(value)){
+                return last.getValue();
+            }else{
+                while (last.getValue() != null){
+                    if(last.getValue().equals(value)){
+                        return last.getValue();
+                    }
+                    last = last.getPrev();
+                }
+            }
```
Comment (Fabolo12) [inline src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java]: Doesn't work with boundary values and not existing values

File: src/test/java/com/izmaylov/hw15/practiceSlide20/MyListTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Test
+    void shouldIncreaseSizeByOneWhenTryingToAddValue(){
+        MyList myList = new MyList();
+        String temp = "1";
+        int expected = 1;
+        myList.add(temp);
+
+
+        Assert.assertEquals(expected, myList.size);
```
Comment (Fabolo12) [inline src/test/java/com/izmaylov/hw15/practiceSlide20/MyListTest.java]: If you are using the 5th version of the library, then you must use the checks from it: Assertions

File: src/test/java/com/izmaylov/hw15/practiceSlide33/BoxComparatorTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
+package com.izmaylov.hw15.practiceSlide33;
+
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class BoxComparatorTest {
```
Comment (Fabolo12) [inline src/test/java/com/izmaylov/hw15/practiceSlide33/BoxComparatorTest.java:8]: I would add tests to check 0 as it has special behavior

**Review by Fabolo12 — CHANGES_REQUESTED** [review]

File: src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    private MyNode<E> tail;
+    private int size = 0;
+
+    public int getSize() {
+        return size;
+    }
+
+    public void add(E value) {
+        if (value == null) {
+            throw new NullPointerException();
```
Comment (Fabolo12) [inline src/main/java/com/izmaylov/hw15/practiceSlide20/MyList.java]: Better IllegalArgumentException

File: src/test/java/com/izmaylov/hw15/practiceSlide20/MyListTest.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        Assertions.assertEquals(expected, myList.getSize());
+
+    }
+
+    @Test
+    void shouldThrowExceptionBecauseLastValueIsNull(){
+        MyList myList = new MyList();
+
+        Assertions.assertThrows(NoSuchElementException.class,() -> myList.getFromLast(3));
+    }
```
Comment (Fabolo12) [inline src/test/java/com/izmaylov/hw15/practiceSlide20/MyListTest.java]: Now this test fails :)

**Review by Fabolo12 — APPROVED** [review]
10
